File: src/main/java/com/windchopper/fs/sftp/SftpFileSystemProvider.java
package com.windchopper.fs.sftp;

import com.jcraft.jsch.JSch;

import java.io.IOException;
import java.net.URI;
import java.nio.channels.SeekableByteChannel;
import java.nio.file.*;
import java.nio.file.DirectoryStream.Filter;
import java.nio.file.attribute.BasicFileAttributes;
import java.nio.file.attribute.FileAttribute;
import java.nio.file.attribute.FileAttributeView;
import java.nio.file.spi.FileSystemProvider;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import static java.util.Collections.emptyMap;

public class SftpFileSystemProvider extends FileSystemProvider implements SftpConstants, SftpRoutines {

    final Map<SftpSessionIdentity, SftpFileSystem> connectedFileSystems = new ConcurrentHashMap<>();
    final JSch jsch = new JSch();

    @Override public String getScheme() {
        return SftpConfiguration.SCHEME;
    }

    @Override public FileSystem newFileSystem(URI uri, Map<String, ?> environment) throws IOException {
        SftpConfiguration configuration = new SftpConfiguration(uri, environment);

        if (connectedFileSystems.containsKey(configuration.sessionIdentity)) {
            throw new FileSystemAlreadyExistsException(
                String.format("File system already connected to %s", configuration.sessionIdentity));
        }

        SftpFileSystem fileSystem = new SftpFileSystem(this, configuration);
        connectedFileSystems.put(configuration.sessionIdentity, fileSystem);
        return fileSystem;
    }

    @Override public FileSystem getFileSystem(URI uri) {
        SftpConfiguration configuration = new SftpConfiguration(uri, emptyMap());

        if (connectedFileSystems.containsKey(configuration.sessionIdentity)) {
            return connectedFileSystems.get(configuration.sessionIdentity);
        }

        throw new FileSystemNotFoundException(
            String.format("File system not connected to %s", configuration.sessionIdentity));
    }

    void retire(SftpSessionIdentity connectionIdentity, SftpFileSystem fileSystem) {
        connectedFileSystems.remove(connectionIdentity, fileSystem);
    }

    @Override public Path getPath(URI uri) {
        return getFileSystem(uri).getPath(uri.getPath());
    }

    @Override public SeekableByteChannel newByteChannel(Path path, Set<? extends OpenOption> openOptionSet, FileAttribute<?>... fileAttributes) throws IOException {
        return checkPathAndApply(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public DirectoryStream<Path> newDirectoryStream(Path path, Filter<? super Path> filter) throws IOException {
        return checkPathAndApply(path, domesticPath -> domesticPath.getFileSystem().newDirectoryStream(domesticPath, filter));
    }

    @Override public void createDirectory(Path path, FileAttribute<?>... fileAttributes) throws IOException {
        checkPathAndAccept(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public void delete(Path path) throws IOException {
        checkPathAndAccept(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    void delete(SftpPath path) throws IOException {
        throw new UnsupportedOperationException();
    }

    @Override public void copy(Path sourcePath, Path targetPath, CopyOption... copyOptions) throws IOException {
        checkPathAndAccept(targetPath, domesticTargetPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public void move(Path sourcePath, Path targetPath, CopyOption... copyOptions) throws IOException {
        checkPathAndAccept(targetPath, domesticTargetPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public boolean isSameFile(Path path1st, Path path2nd) throws IOException {
        return checkPathAndApply(path1st, path2nd, (domesticPath1st, domesticPath2nd) -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public boolean isHidden(Path path) throws IOException {
        return checkPathAndApply(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public FileStore getFileStore(Path path) throws IOException {
        return checkPathAndApply(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public void checkAccess(Path path, AccessMode... accessModes) throws IOException {
        checkPathAndAccept(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public <V extends FileAttributeView> V getFileAttributeView(Path path, Class<V> viewType, LinkOption... linkOptions) {
        return checkPathAndApply(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @SuppressWarnings("unchecked") @Override public <A extends BasicFileAttributes> A readAttributes(Path path, Class<A> attributesType, LinkOption... linkOptions) throws IOException {
        return checkPathAndApply(path, domesticPath -> {
            if (attributesType == BasicFileAttributes.class || attributesType == SftpFileAttributes.class) {
                return (A) domesticPath.getFileSystem().view(path.toString()).toFileAttributes();
            }

            throw new UnsupportedOperationException(
                String.format("Attributes of type %s not supported", attributesType.getCanonicalName()));
        });
    }

    @Override public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... linkOptions) throws IOException {
        return checkPathAndApply(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

    @Override public void setAttribute(Path path, String s, Object value, LinkOption... linkOptions) throws IOException {
        checkPathAndAccept(path, domesticPath -> {
            throw new UnsupportedOperationException();
        });
    }

}

File: src/main/java/com/github/windchopper/fs/sftp/SftpFileSystemProvider.kt
package com.github.windchopper.fs.sftp

import com.github.windchopper.fs.JSchLoggerBridge
import com.jcraft.jsch.JSch
import java.io.IOException
import java.net.URI
import java.nio.channels.SeekableByteChannel
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.nio.file.attribute.FileAttribute
import java.nio.file.attribute.FileAttributeView
import java.nio.file.spi.FileSystemProvider
import java.util.concurrent.ConcurrentHashMap
import java.util.logging.Logger

class SftpFileSystemProvider: FileSystemProvider(), SftpRoutines {

    val secureChannel = JSch()
    val connectedFileSystems: MutableMap<SftpConfiguration.SessionIdentity, SftpFileSystem> = ConcurrentHashMap()

    init {
        JSch.setLogger(JSchLoggerBridge(Logger.getLogger(this::class.qualifiedName)))

        System.getProperty("user.home")
            ?.let { Paths.get(it).resolve(".ssh") }
            ?.let {
                wrapNotIOException {
                    listOf(it.resolve("id_rsa"), it.resolve("id_dsa"), it.resolve("id_ecdsa"))
                        .filter { Files.exists(it) }
                        .forEach { secureChannel.addIdentity(it.toRealPath().toString()) }

                    val knownHostsFile = it.resolve("known_hosts")

                    if (Files.exists(knownHostsFile)) {
                        secureChannel.setKnownHosts(knownHostsFile.toRealPath().toString())
                    }
                }
            }
    }

    override fun getScheme(): String {
        return SftpConstants.SCHEME
    }

    override fun newFileSystem(uri: URI, environment: Map<String, *>): FileSystem {
        with (SftpConfiguration(uri, environment)) {
            if (connectedFileSystems.containsKey(sessionIdentity)) {
                throw FileSystemAlreadyExistsException("File system already connected to ${sessionIdentity}")
            }

            return SftpFileSystem(this@SftpFileSystemProvider, this)
                .let {
                    connectedFileSystems[sessionIdentity] = it
                    it
                }
        }
    }

    override fun getFileSystem(uri: URI): FileSystem {
        with (SftpConfiguration(uri)) {
            if (connectedFileSystems.containsKey(sessionIdentity)) {
                return connectedFileSystems[sessionIdentity]!!
            }

            throw FileSystemNotFoundException("File system not connected to ${sessionIdentity}")
        }
    }

    fun retire(connectionIdentity: SftpConfiguration.SessionIdentity, fileSystem: SftpFileSystem?) {
        connectedFileSystems.remove(connectionIdentity, fileSystem)
    }

    override fun getPath(uri: URI): Path {
        return getFileSystem(uri).getPath(uri.path)
    }

    @Throws(IOException::class) override fun newByteChannel(path: Path, openOptionSet: Set<OpenOption?>, vararg fileAttributes: FileAttribute<*>?): SeekableByteChannel {
        return checkPathAndApply<SeekableByteChannel>(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) override fun newDirectoryStream(path: Path, filter: DirectoryStream.Filter<in Path?>): DirectoryStream<Path> {
        return checkPathAndApply(path) { it.fileSystem.newDirectoryStream(it, filter) }
    }

    @Throws(IOException::class) override fun createDirectory(path: Path, vararg fileAttributes: FileAttribute<*>?) {
        checkPathAndAccept(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) override fun delete(path: Path) {
        checkPathAndAccept(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) fun delete(path: SftpPath?) {
        TODO("not implemented")
    }

    @Throws(IOException::class) override fun copy(sourcePath: Path, targetPath: Path, vararg copyOptions: CopyOption) {
        checkPathAndAccept(targetPath) { TODO("not implemented") }
    }

    @Throws(IOException::class) override fun move(sourcePath: Path, targetPath: Path, vararg copyOptions: CopyOption) {
        checkPathAndAccept(targetPath) { TODO("not implemented") }
    }

    @Throws(IOException::class) @Suppress("NAME_SHADOWING") override fun isSameFile(path1st: Path, path2nd: Path): Boolean {
        return checkPathAndApply(path1st, path2nd) { path1st, path2nd -> TODO("not implemented") }
    }

    @Throws(IOException::class) override fun isHidden(path: Path): Boolean {
        return checkPathAndApply<Boolean>(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) override fun getFileStore(path: Path): FileStore {
        return checkPathAndApply<FileStore>(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) override fun checkAccess(path: Path, vararg accessModes: AccessMode) {
        checkPathAndAccept(path) { TODO("not implemented") }
    }

    override fun <V: FileAttributeView?> getFileAttributeView(path: Path, viewType: Class<V>, vararg linkOptions: LinkOption): V {
        return checkPathAndApply(path) { TODO("not implemented") }
    }

    @Throws(IOException::class) @Suppress("UNCHECKED_CAST") override fun <A: BasicFileAttributes> readAttributes(path: Path, attributesType: Class<A>, vararg linkOptions: LinkOption): A {
        return checkPathAndApply(path) {
            return@checkPathAndApply if (attributesType === BasicFileAttributes::javaClass || attributesType === SftpFileAttributes::javaClass) {
                it.fileSystem.view(path.toString()).toFileAttributes() as A
            } else {
                throw attributesNotSupported(attributesType)
            }
        }
    }

    override fun readAttributes(path: Path, attributes: String, vararg linkOptions: LinkOption): Map<String, Any> {
        return checkPathAndApply(path) { TODO("not implemented") }
    }

    override fun setAttribute(path: Path, s: String, value: Any, vararg linkOptions: LinkOption) {
        checkPathAndAccept(path) { TODO("not implemented") }
    }

}